05-AdvancedTopics/web-search-mcp/README.md

```diff
@@ -0,0 +1,214 @@
+The server will run as a stdio-based MCP server that the client can connect to directly.
+
+### Running the Client
+
+Run the automated tests (this will automatically start the server):
+
+```bash
+python client.py
+```
+
+Or run in interactive mode:
+
+```bash
+python client.py --interactive
+```
+
+### Testing with Different Methods
+
+#### Using the Interactive Mode
+The easiest way to test the tools is using the interactive mode:
+
+```bash
+python client.py --interactive
+```
+
+This provides a menu-driven interface to try each tool with custom queries.
+
+#### Using the MCP Python SDK
+You can also build your own test scripts using the MCP Python SDK:
+
+```python
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.stdio import stdio_client
+
+async def test_custom_query():
+    server_params = StdioServerParameters(
+        command="python",
+        args=["server.py"],
+    )
+    
+    async with stdio_client(server_params) as (reader, writer):
+        async with ClientSession(reader, writer) as session:
+            await session.initialize()
+            # Call tools with your custom parameters
+            result = await session.call_tool("general_search", 
+                                           arguments={"query": "your custom query"})
+            # Process the result
+```
+
+## Tool Descriptions
+
+### general_search
+
+Performs a general web search and returns formatted results.
+
+**Parameters:**
+- `query` (string): The search query
+- `num_results` (integer, optional): Number of results to return (default: 5)
+
+**Example Request:**
+```json
+{
+  "query": "latest AI trends"
+}
+```
+
+### news_search
+
+Searches for recent news articles related to a query.
+
+**Parameters:**
+- `query` (string): The search query
+- `num_results` (integer, optional): Number of news articles to return (default: 5)
+
+**Example Request:**
+```json
+{
+  "query": "AI policy updates"
+}
+```
+
+### product_search
+
+Searches for products matching a query.
+
+**Parameters:**
+- `query` (string): The product search query
+- `num_results` (integer, optional): Number of product results to return (default: 5)
+
+**Example Request:**
+```json
+{
+  "query": "best AI gadgets 2025"
+}
+```
+
+### qna
+
+Gets direct answers to questions from search engines.
+
+**Parameters:**
+- `question` (string): The question to find an answer for
+
+**Example Request:**
+```json
+{
+  "question": "what is artificial intelligence"
+}
+```
+
+## Advanced Concepts Demonstrated
+
+1. **Multi-tool Orchestration**: Managing multiple tools with distinct purposes within a single server
+2. **API Rate Limit Handling**: Built-in timeout handling for external API calls
+3. **Structured Data Parsing**: Converting complex API responses into formatted, LLM-friendly text
+4. **Error Recovery**: Graceful degradation when API calls fail
+5. **Parameter Validation**: Proper handling of optional parameters with defaults
+
+## Troubleshooting
+
+### Common Issues
```

Comment:
Explain this using real sentences feels very short. It's a good section though as I'm sure the reader appreciates being told this information